python/lsst/daf/butler/script/rewriteSqliteRegistry.py

```diff
@@ -0,0 +1,202 @@
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+__all__ = ("rewriteSqliteRegistry",)
+
+import io
+import logging
+from collections import defaultdict
+from typing import Dict
+
+from lsst.daf.butler import Config, Butler, DatasetAssociation, DatasetIdGenEnum, DatasetRef, DatasetId
+from lsst.daf.butler.core.utils import getClassOf
+from lsst.daf.butler.transfers import RepoExportContext
+from lsst.daf.butler.registry import CollectionType
+
+log = logging.getLogger(__name__)
+
+
+def rewriteSqliteRegistry(source: str, dest: str) -> None:
+    """Rewrite a SQLite registry as a new registry.
+
+    * Create a new sqlite registry based on the current configuration.
+    * Export all dimension records, dataset types, collection information
+      and import into new registry.
+    * Transfer datasets without moving them.
+
+    Question is whether we get clever and try to move the old registry
+    out of the way and replace it along with a modified butler.yaml making
+    it look like an inplace rewrite.
+
+    """
+
+    # Create the source butler early so we can ask it questions
+    # without assuming things.
+    source_butler = Butler(source, writeable=False)
+
+    # The source butler knows where its config came from.
+    source_config_uri = source_butler._config.configFile
+
+    # We need to read just the local configuration and not the expanded
+    # one.
+    config = Config(source_config_uri)
+
+    # Assume that we are rewriting this with the current set of registry
+    # managers so remove any.
+    del config["registry", "managers"]
+
+    # Delete the URI of the registry since we want an entirely new
+    # entry.
+    del config["registry", "db"]
+
+    # Force the new temporary butler repo to have an explicit path
+    # to the existing datastore.  This will only work for a FileDatastore
+    # so try that and if it doesn't work warn but continue since either
+    # this is a new typoe of datastore or a chained datastore of some kind.
+    # For now only handle the simple case.
+    # NOTE: Execution Butler creation has a similar problem with working
+    # out how to refer back to the original datastore.
+    try:
+        config["datastore", "root"] = str(source_butler.datastore.root)  # type: ignore
+
+        # Force the name of the datastore since that should not
+        # change from the source (and will if set from root)
+        config["datastore", "name"] = source_butler.datastore.name
+    except AttributeError:
```

Comment:
Could you just check the type of `butler.datastore`? Relying on missing attributes may not be super-reliable.